Closing a Netlink socket with ongoing Receive() call

Hi Matt

I'm trying to cleanly terminate a Netlink socket listening on a multicast group.

```go
func main() {
	conn, err := netlink.Dial(0xc, nil) // NETLINK_NETFILTER
	if err != nil {
		log.Fatal(err)
	}

	err = conn.JoinGroup(10) // Non-existent multicast group (Receive() will never return)
	if err != nil {
		log.Fatal(err)
	}

	recvChan := make(chan []netlink.Message)

	// Start a blocking receive
	go func() {
		recv, err := conn.Receive()
		if err != nil {
			log.Fatal(err)
		}
		recvChan <- recv
	}()

	// Make sure goroutine enters Receive()
	// time.Sleep(50 * time.Millisecond)

	err = conn.Close()
	if err != nil {
		log.Fatal(err)
	}

	log.Println(<-recvChan)
}
```

This uncovers two issues:

1. A race condition: running this program multiple times in sequence yields different results. One is a panic:

```
panic: send on closed channel
goroutine 4 [running]:
github.com/mdlayher/netlink.(*sysSocket).do(0xc4200bc0a0, 0xc4200f6000)
    /home/timo/go/src/github.com/mdlayher/netlink/conn_linux.go:356 +0xa8
```

.. that can be avoided by closing the channel after the WaitGroup merges. This causes a deadlock in the test suite, so this might invalidate some assumptions you've made in other places.

```go
func (s *sysSocket) Close() error {
	...
	s.wg.Wait()
	close(s.funcC)
```

2. `Receive()` is non-interruptible. (uncomment the `Sleep()` in the example)

If there's no valid Netlink message coming in (like in the example above), `Receive()` will hang forever. I've tried closing the fd from the main thread by executing `unix.Close()` directly instead of queueing it with `s.do()`, but this (expectedly) doesn't interrupt the syscall either. It will only yield when a message is read from the socket.

I know this is not directly an issue with the library itself, but it would be pleasant to have a solution or workaround. Is there anything you could think of? A quick search hinted that syscalls can be interrupted by signals, not sure if this is something we could implement.

--- a/conn_linux.go
+++ b/conn_linux.go
@@ -299,6 +299,9 @@ type sysSocket struct {
 
 	wg    *sync.WaitGroup
 	funcC chan<- func()
+
+	mu   sync.RWMutex
+	done bool
 }
 
 // newSysSocket creates a sysSocket that optionally locks its internal goroutine
@@ -361,6 +364,12 @@ func (s *sysSocket) do(f func()) {
 }
 
 func (s *sysSocket) Socket(family int) error {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	if s.done {
+		return syscall.EBADF
+	}
+
 	var (
 		fd  int
 		err error
@@ -382,6 +391,12 @@ func (s *sysSocket) Socket(family int) error {
 }
 
 func (s *sysSocket) Bind(sa unix.Sockaddr) error {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	if s.done {
+		return syscall.EBADF
+	}
+
 	var err error
 	s.do(func() {
 		err = unix.Bind(s.fd, sa)
@@ -391,11 +406,23 @@ func (s *sysSocket) Bind(sa unix.Sockaddr) error {
 }
 
 func (s *sysSocket) Close() error {
+	// Be sure to acquire a write lock because we need to stop any other
+	// goroutines from sending system call requests after close.
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if s.done {
+		return syscall.EBADF
+	}
+
 	var err error
 	s.do(func() {
 		err = unix.Close(s.fd)
 	})
 
+	// No more requests should be processed.  Terminate the system call
+	// handling goroutine by closing the channel it is listening on,
+	// and lock out any future requests to avoid send on closed channel panics.
+	s.done = true
 	close(s.funcC)
 	s.wg.Wait()
 
@@ -405,6 +432,12 @@ func (s *sysSocket) Close() error {
 func (s *sysSocket) FD() int { return s.fd }
 
 func (s *sysSocket) Getsockname() (unix.Sockaddr, error) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	if s.done {
+		return nil, syscall.EBADF
+	}
+
 	var (
 		sa  unix.Sockaddr
 		err error
@@ -416,7 +449,14 @@ func (s *sysSocket) Getsockname() (unix.Sockaddr, error) {
 
 	return sa, err
 }
+
 func (s *sysSocket) Recvmsg(p, oob []byte, flags int) (int, int, int, unix.Sockaddr, error) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	if s.done {
+		return 0, 0, 0, nil, syscall.EBADF
+	}
+
 	var (
 		n, oobn, recvflags int
 		from               unix.Sockaddr
@@ -431,6 +471,12 @@ func (s *sysSocket) Recvmsg(p, oob []byte, flags int) (int, int, int, unix.Socka
 }
 
 func (s *sysSocket) Sendmsg(p, oob []byte, to unix.Sockaddr, flags int) error {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	if s.done {
+		return syscall.EBADF
+	}
+
 	var err error
 	s.do(func() {
 		err = unix.Sendmsg(s.fd, p, oob, to, flags)
@@ -440,6 +486,12 @@ func (s *sysSocket) Sendmsg(p, oob []byte, to unix.Sockaddr, flags int) error {
 }
 
 func (s *sysSocket) SetSockopt(level, name int, v unsafe.Pointer, l uint32) error {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	if s.done {
+		return syscall.EBADF
+	}
+
 	var err error
 	s.do(func() {
 		err = setsockopt(s.fd, level, name, v, l)

--- a/conn_linux_test.go
+++ b/conn_linux_test.go
@@ -12,6 +12,7 @@ import (
 	"testing"
 	"unsafe"
 
+	"github.com/google/go-cmp/cmp"
 	"github.com/mdlayher/netlink/nlenc"
 	"golang.org/x/sys/unix"
 )
@@ -422,6 +423,25 @@ func TestLinuxConnIntegrationConcurrent(t *testing.T) {
 	}
 
 	wg.Wait()
+}
+
+func TestLinuxConnIntegrationClosedConn(t *testing.T) {
+	const familyGeneric = 16
+
+	c, err := Dial(familyGeneric, nil)
+	if err != nil {
+		t.Fatalf("failed to dial netlink: %v", err)
+	}
+
+	// Close the connection immediately and ensure that future calls get EBADF.
+	if err := c.Close(); err != nil {
+		t.Fatalf("failed to close: %v", err)
+	}
+
+	_, err = c.Receive()
+	if diff := cmp.Diff(syscall.EBADF, err); diff != "" {
+		t.Fatalf("unexpected error  (-want +got):\n%s", diff)
+	}
 }
 
 func TestLinuxConnJoinLeaveGroup(t *testing.T) {

